packages/hardhat/contracts/BitPiq.sol

```diff
@@ -0,0 +1,133 @@
+//SPDX-License-Identifier: MIT
+pragma solidity >=0.8.0 <0.9.0;
+
+// Useful for debugging. Remove when deploying to a live network.
+import "hardhat/console.sol";
+
+/**
+ * A smart contract that allows users to bet on the last 4 bits of a block hash
+ * Users can place bets by specifying a 4-bit hash, block number, and amount of ETH
+ * Once the target block is mined, users can claim winnings if their hash matches
+ */
+contract BitPiq {
+    address public immutable owner;
+    uint256 public winningsReserve;
+
+    struct Bet {
+        uint8 hashPick;
+        uint256 blockNumber;
+        uint256 ethAmount;
+        uint256 betId;
+        bool claimed;
+    }
+
+    mapping(address => mapping(uint256 => Bet)) public Bets;
+    mapping(address => uint256) public NextBetId;
+
+    event BetPlaced(address indexed user, uint256 betId, uint8 hashPick, uint256 ethAmount, uint256 blockNumber);
+
+    constructor() {
+        owner = msg.sender;
+    }
+
+    modifier isOwner() {
+        require(msg.sender == owner, "Not the Owner");
+        _;
+    }
+
+    /**
+     * @notice Places a bet on the last 4 bits of a future block hash
+     * @param _hashPick uint8 - 4-bit number to bet on (0-15)
+     */
+    function placeBet(uint8 _hashPick) public payable {
+        require(_hashPick <= 15, "Hash pick must be 4 bits (0-15)");
+        require(address(this).balance >= (msg.value * 16) + winningsReserve, "Insufficient amount in reserves");
+
+        uint256 nextBetId = NextBetId[msg.sender];
+
+        Bets[msg.sender][nextBetId] = Bet({
+            hashPick: _hashPick,
+            blockNumber: block.number,
+            ethAmount: msg.value,
+            betId: nextBetId,
+            claimed: false
+        });
+
+        emit BetPlaced(msg.sender, nextBetId, _hashPick, msg.value, block.number);
+
+        NextBetId[msg.sender]++;
+        winningsReserve += msg.value * 16;
+    }
+
+    /**
+     * @notice Function for user to claim winnings for multiple bets
+     * @param _address The user's address
+     * @param _betIds Array of bet IDs to claim winnings for
+     */
+    function claimWinnings(address _address, uint256[] calldata _betIds) public {
+        require(_betIds.length > 0, "_betIds array must not be empty!");
+
+        uint256 totalWinnings = 0;
+
+        for (uint256 i = 0; i < _betIds.length; i++) {
+            uint256 betId = _betIds[i];
+            Bet storage bet = Bets[_address][betId];
+
+            require(bet.blockNumber != 0, "Input includes an invalid bet id!");
+            require(!bet.claimed, "Input includes an already claimed bet!");
```

Comment:
I think that this could cause a bug. Does this exit the function if the condition isn't met? If so, we could potentially mark certain bets as claimed, then reach an invalid betId and exit the function without ever paying out the total winnings.

Reply:
I believe that a function call is an atomic transaction, and that all state changes that occurred are rolled back if a require throws an error. It either succeeds entirely or fails completely.

Reply:
But i appreciate the call out. Early on I just flipped that assumption in my head and didn't even really think about it. Without the atomic transaction behavior, this would indeed be a bug